### PR TITLE
Fix download programming repositories

### PR DESCRIPTION
**Pull request by FelixTJDietrich**
### Motivation and Context

@bassner had some issues running the script and I'm addressing those issues in this PR.

### Description

- Use `axios` for requests now (this probably also lowers the node version requirement to node 16)
- Improve general error handling/logging
- Remove parallelization in favor of a more long running script that is more transparent with what is going on
- Add participation ids to the evaluation data 
- Download the repositories in batches of 25 instead of having one request downloading potentially ~1000 repos at a time 
- Write `download-<timestamp>.txt` and `submission-<timestamp>.txt` for logging downloaded/failed exercise/participation ids

### Steps for Testing

1. `cd playground`
2. `npm install`
3. Use the data from below in `data/evaluation/programming-exercises.json`
4. Run `npm run export:artemis:3-download-programming-repositories`
     a. Login with instructor test account on Artemis (production)
     b. Select all exercises
     c. Download
5. It should now run in batches

```
{
  "TEST": {
    "semester": "WS20/21",
    "course_id": 86,
    "exercises": [
      {
        "id": 10573,
        "title": "TEST 01",
        "participations": [
          7945271,
          8082848,
          7936574
        ],
        "is_exam_exercise": 0
      },
      {
        "id": 10573,
        "title": "TEST 02",
        "participations": [
          7945271,
          8082848,
          7936574
        ],
        "is_exam_exercise": 0
      }
    ]
  }
}
```

### Screenshots

image_720

### CODE REVIEW

File: playground/scripts/artemis/3_download_programming_repositories.mjs [hunk 1]
```diff
@@ -1,70 +1,39 @@
...
-
-let baseURL = "";
-let authCookie = "";
+import { programming, evaluationOutputDirPath } from "./utils.mjs";
 
 // Number of times to retry a fetch if it fails
-const fetchRetryCount = 3;
-// Delay between fetch retries in milliseconds
+const fetchRetries = 4;
 const fetchRetryDelay = 1000;
```
Comment (pal03377) [inline playground/scripts/artemis/3_download_programming_repositories.mjs]: I think the comment is still valuable for the `fetchRetryDelay` line.
```suggestion
const fetchRetries = 4;
// Delay between fetch retries in milliseconds
const fetchRetryDelay = 1000;
```

File: playground/scripts/artemis/3_download_programming_repositories.mjs [hunk 2]
```diff
@@ -1,70 +1,39 @@
...
-const exponentialBackoffFactor = 5; // 1, 5, 25 (seconds)
-const concurrentDownloads = 5;
+const timeout = 60 * 1000 * 10; // 10 minutes
```
Comment (pal03377) [inline playground/scripts/artemis/3_download_programming_repositories.mjs]:
```suggestion
const timeoutMilliseconds = 10 * 60 * 1000; // 10 minutes
```
optional

File: playground/scripts/artemis/3_download_programming_repositories.mjs [hunk 3]
```diff
@@ -78,33 +47,97 @@ async function auth() {
...
     console.log("Authenticated successfully");
-  } else {
-    console.error("Failed to authenticate: No cookie received");
+    artemisInstance = axiosInstance;
+  } catch (error) {
+    console.error(
+      "An error occurred during authentication:",
+      JSON.stringify(error)
+    );
     process.exit(1);
```
Comment (pal03377) [inline playground/scripts/artemis/3_download_programming_repositories.mjs]:
```suggestion
    console.error("An error occurred during authentication:");
    throw error;
```
optional, would log the stack trace in a nicer format

File: playground/scripts/artemis/3_download_programming_repositories.mjs [hunk 4]
```diff
@@ -199,143 +233,151 @@ async function downloadMaterial(exerciseId) {
...
-    );
-    return false;
-  }
+async function download(exercise) {
+  let success = true;
+  console.log(`Downloading exercise ${exercise.id}...`);
+  success = success && (await downloadMaterial(exercise.id));
+  success =
+    success &&
+    (await downloadSubmissions(exercise.id, exercise.participations));
```
Comment (pal03377) [inline playground/scripts/artemis/3_download_programming_repositories.mjs]: I don't get these changes. Was this in a loop at some point? The first line is a bit redundantly written if I'm not missing something... Also, imo this logic relies a bit too much on short circuiting.
```suggestion
  success = await downloadMaterial(exercise.id);
  if (success) {
      // both downloads have to be successful
      success = await downloadSubmissions(exercise.id, exercise.participations);
  }
```

**Review by pal03377 — APPROVED** [review]
The code looks good! I have a few suggestions below. Feel free to apply them as you wish / think is useful and then merge the PR.

File: playground/scripts/artemis/3_download_programming_repositories.mjs (hunk 2 above)
Comment (FelixTJDietrich) [reply playground/scripts/artemis/3_download_programming_repositories.mjs]: Added

**Pull request merged**